### ISSUE TITLE
fuzz: null-deref when  WRITE memory access　

### ISSUE THREAD

**Issue by eaglegai** [haproxy/haproxy#2178]
### Detailed Description of the Problem

I'm doing fuzz test recently for haproxy and there is  a null-deref when WRITE memory access.

```
/out/haproxy/fuzz_cfg_parser: Running 1 inputs 1 time(s) each.
[ALERT]    (694) : parsing [/tmp/libfuzzer.694:1]: unknown keyword '?$?????????i.[' out of section.
[ALERT]    (694) : parsing [/tmp/libfuzzer.694:2]: unknown keyword 'u' out of section.

Thread 1 "fuzz_cfg_parser" received signal SIGSEGV, Segmentation fault.
make_arg_list (in=<optimized out>, len=0, mask=2, argp=0x6030000001c8, err_msg=<optimized out>, end_ptr=<optimized out>, err_arg=<optimized out>, al=<optimized out>) at src/arg.c:228
228                     trash.area[trash.data] = 0;
(gdb) bt
#0  make_arg_list (in=<optimized out>, len=0, mask=2, argp=0x6030000001c8, err_msg=<optimized out>, end_ptr=<optimized out>, err_arg=<optimized out>, al=<optimized out>) at src/arg.c:228
#1  0x0000000000809da3 in cfg_parse_cond_term (text=<optimized out>, term=<optimized out>, err=<optimized out>, errptr=<optimized out>, maxdepth=<optimized out>) at src/cfgcond.c:153
#2  0x000000000080b662 in cfg_parse_cond_and (text=0xffffefe3904, expr=<optimized out>, err=<optimized out>, errptr=<optimized out>, maxdepth=<optimized out>) at src/cfgcond.c:336
#3  0x000000000080a522 in cfg_parse_cond_expr (text=0x7ffff7f6cd60, expr=<optimized out>, err=<optimized out>, errptr=<optimized out>, maxdepth=<optimized out>) at src/cfgcond.c:409
#4  0x000000000080bedc in cfg_eval_condition (args=<optimized out>, err=<optimized out>, errptr=<optimized out>) at src/cfgcond.c:494
#5  0x000000000055f22c in readcfgfile (file=<optimized out>) at src/cfgparse.c:1980
#6  0x00000000005545fc in LLVMFuzzerTestOneInput (data=<optimized out>, size=<optimized out>) at fuzz_cfg_parser.c:42
#7  0x000000000045a8f2 in ExecuteCallback () at /src/llvm/projects/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:556
#8  0x0000000000445512 in RunOneTest () at /src/llvm/projects/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:292
#9  0x000000000044b1cf in FuzzerDriver () at /src/llvm/projects/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:774
#10 0x0000000000475343 in main () at /src/llvm/projects/compiler-rt/lib/fuzzer/FuzzerMain.cpp:19
(gdb) f 0
#0  make_arg_list (in=<optimized out>, len=0, mask=2, argp=0x6030000001c8, err_msg=<optimized out>, end_ptr=<optimized out>, err_arg=<optimized out>, al=<optimized out>) at src/arg.c:228
228                     trash.area[trash.data] = 0;
(gdb) p trash
$1 = {size = 0, area = 0x0, data = 0, head = 0}
(gdb)

```

### Expected Behavior

none null-deref 

### Steps to Reproduce the Behavior

see bt

### Do you have any idea what may have caused this?

trash.area is NULL 

### Do you have an idea how to solve the issue?

```
From f84a9efa19adc1ca8f899e1c03e36778bd54e44e Mon Sep 17 00:00:00 2001
From: eaglegai <eaglegai@163.com>
Date: Thu, 8 Jun 2023 20:24:57 +0800
Subject: [PATCH] fuzz : arg: fix a null-deref when a WRITE memory access

make_arg_list use trash  but it doesn't check if trash.area is NULL，then
'trash.area[trash.data] = 0'   or ' *out++ ='  will cause null-deref 

---
 src/arg.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/arg.c b/src/arg.c
index 039602a..cfcfa55 100644
--- a/src/arg.c
+++ b/src/arg.c
@@ -162,6 +162,8 @@ int make_arg_list(const char *in, int len, uint64_t mask, struct arg **argp,
 
 		chunk_reset(&trash);
 		out = trash.area;
+		if (!out)
+			goto empty_err;
 
 		while (len && *in && trash.data < trash.size - 1) {
 			if (*in == '"' && !squote) {  /* double quote outside single quotes */
-- 
2.33.0


```

### What is your configuration?

```haproxy
NA
```

### Output of `haproxy -vv`

```plain
2.6.6  and  the lastest version have the same problem.
```

### Last Outputs and Backtraces

```plain
see above
```

**Comment by eaglegai** [comment]
@wtarreau hi, could you help check this issue?

**Comment by Darlelet** [comment]
I don't see how this is possible, trash is guaranteed to be allocated through `init_trash_buffers()` before `readcfgfile()` is first called (verified for 2.6 and dev). Perhaps this could be the result of another bug which corrupts trash struct during config parsing?

Given that the source code of `readcfgfile()` hasn't changed in a while, I suspect that this could be related to #1507 and https://vulners.com/ossfuzz/OSSFUZZ-23653

**Comment by Darlelet** [comment]
@eaglegai do you know if you can isolate the config (/payload) that triggers this?
Thanks

**Comment by eaglegai** [comment]
> 

sorry, it just be triggered in fuzz testcase，the case is https://github.com/google/oss-fuzz/blob/master/projects/haproxy/fuzz_cfg_parser.c

**Comment by eaglegai** [comment]
I have read the code，trash.area couldn't be used by  readcfgfile() when it is NULL, it will exit from here:
```
        if (!init_trash_buffers(1)) {
                 ha_alert("failed to initialize trash buffers.\n");
                 exit(1);
```

**Comment by Darlelet** [comment]
Thanks, then I guess when you notice a crash you might be able to provide the latest file that was tested, ie: `/tmp/libfuzzer.694` file from your initial report?

This could help reproduce to isolate the issue since deploying the tool and making multiple fuzz attempts is not very convenient and a bit random.

**Comment by Darlelet** [comment]
>I have read the code，trash.area couldnt be used by readcfgfile() when it is NULL, it will exit from here:

Yes indeed, it is likely a memory corruption when parsing the conf that has unexpected side-effects.

**Comment by eaglegai** [comment]
> > I have read the code，trash.area couldnt be used by readcfgfile() when it is NULL, it will exit from here:
> 
> Yes indeed, it is likely a memory corruption when parsing the conf that has unexpected side-effects.

I can't  understand what you mean well,  I mean if trash.area allocation failed, haproxy will exit, readcfgfile() won‘t be executed. the protect code( like patch above )is not  necessary.  you think it's better to protect for  memory corruption？

**Comment by Darlelet** [comment]
There might be some possibility during conf parsing (in `readcfgfile()`) that a bad input triggers a bug under `readcfgfile()` (ie: stack overflow or buffer overflow) that doesn't happen with sane input, and because of this haproxy instead of crashing immediately keeps running for a while in an undefined behavior state that would eventually trigger a crash. In your case this results in `trash.area` becoming NULL but this is likely the consequence of the previous buffer or stack overflow.

Due to this, in your gdb trace the process memory cannot be trusted since it was probably corrupted some time before the process actually crashed, we need to find the root cause of this memory corruption to dig further :)

Thus a reproducer (bad config file) would be very helpful to try to isolate the config bits that trigger this.

**Comment by eaglegai** [comment]
> 

I'm sorry to say the environment has been cleaned up.

**Comment by Darlelet** [comment]
Oh ok :/ nevermind, since you provided the fuzzing source repo I'll try to reproduce on my end.
I'll come back to you if I'm unable to reproduce

**Comment by eaglegai** [comment]
thanks all ~

**Comment by eaglegai** [comment]
> There might be some possibility during conf parsing (in `readcfgfile()`) that a bad input triggers a bug under `readcfgfile()` (ie: stack overflow or buffer overflow) that doesn't happen with sane input, and because of this haproxy instead of crashing immediately keeps running for a while in an undefined behavior state that would eventually trigger a crash. In your case this results in `trash.area` becoming NULL but this is likely the consequence of the previous buffer or stack overflow.
> 
> Due to this, in your gdb trace the process memory cannot be trusted since it was probably corrupted some time before the process actually crashed, we need to find the root cause of this memory corruption to dig further :)
> 
> Thus a reproducer (bad config file) would be very helpful to try to isolate the config bits that trigger this.

**Comment by eaglegai** [comment]
> There might be some possibility during conf parsing (in `readcfgfile()`) that a bad input triggers a bug under `readcfgfile()` (ie: stack overflow or buffer overflow) that doesn't happen with sane input, and because of this haproxy instead of crashing immediately keeps running for a while in an undefined behavior state that would eventually trigger a crash. In your case this results in `trash.area` becoming NULL but this is likely the consequence of the previous buffer or stack overflow.
> 
> Due to this, in your gdb trace the process memory cannot be trusted since it was probably corrupted some time before the process actually crashed, we need to find the root cause of this memory corruption to dig further :)
> 
> Thus a reproducer (bad config file) would be very helpful to try to isolate the config bits that trigger this.

It suddenly occurred to me that I had added debug code to print trash.area at the starting of readcfgfile\cfg_parse_cond_expr\cfg_parse_cond_and\cfg_parse_cond_term, they are all NULL, I don't know if it's helpful for your analyze.

**Comment by Darlelet** [comment]
Ok so I gave it a try and was able to hit and extract a conf that makes `readcfgfile` crash under the fuzzing binary.

However when trying the same config file with a standalone haproxy I was unable to reproduce (I tried 2.6-2.9).
Given that in #1507 Willy already gave it a try with such a "bad" conf and was unable to reproduce, I started to wonder if it had something to do with how haproxy is used in the testcase.

I stumbled on this file which takes care of manually compiling haproxy from the sources, then make a library out of it to embed it in the fuzzer binary: https://github.com/google/oss-fuzz/blob/master/projects/haproxy/build.sh

From what I can see, there is definitely something wrong.. `main()` function from haproxy will never be called because it is renamed to `main2()` to prevent it from conflicting with the `main()` from the fuzzer when making the .a library file out of haproxy sources.

Indeed when the fuzzer is testing haproxy configs, it directly calls `readcfgfile()` from haproxy (https://github.com/google/oss-fuzz/blob/a6cec35f56e5b86f396f87002678f8c905c97e9e/projects/haproxy/fuzz_cfg_parser.c#L38) but as `main()` from haproxy is never executed, all memory initialization (including trash initialization) performed in `main()` which happen under normal use of haproxy are skipped. This explains why `trash.area` is NULL in your case, because `init_trash_buffers()` is called during startup under the `main()` haproxy function. So my first hypothesis was wrong, and you were actually right :)

The `fuzz_cfg_parser.c` testfile was developed with the assumption that `readcfgfile()` function from haproxy is standalone, but this is not the case actually. In its current state, the fuzz test program for haproxy makes no sense and should be fixed to ensure normal haproxy startup sequence. In my opinion the safest approach to monitor its behavior when given fuzzed input configs file would be to keep haproxy as a standalone binary and execute it in config check mode  ("-c" startup option) using `execv()` from the fuzzer binary.

I suggest you to file a report on https://github.com/google/oss-fuzz if you're interested in doing that.
Meanwhile, I guess that this issue may safely be closed.
